dll/win32/shlwapi/propbag.cpp

```diff
@@ -995,3 +995,197 @@ SHCreatePropertyBagOnProfileSection(
+                    L"Settings",
+                    NULL,
+                    &Flags,
+                    &cbValue) != ERROR_SUCCESS || cbValue < sizeof(Flags))
+    {
+        return E_FAIL;
+    }
+
+    V_UINT(pvari) = ((UINT)(Flags >> 32)) | 0x220;
+    V_VT(pvari) = VT_UINT;
+    return S_OK;
+}
+
+// FIXME
+typedef struct tagUNKNOWN_STREAM_DATA
+{
+    WORD wMagic;
+    WORD aw[6];
+    WORD wSize;
+} UNKNOWN_STREAM_DATA, *PUNKNOWN_STREAM_DATA;
+
+IStream* CDesktopUpgradePropertyBag::_NewStreamFromOld(IStream *pOldStream)
+{
+    UNKNOWN_STREAM_DATA Data;
+    HRESULT hr = pOldStream->Read(&Data, sizeof(Data), NULL);
+    if (FAILED(hr) || Data.wMagic != 28)
+        return NULL;
+
+    // Move stream pointer
+    LARGE_INTEGER li;
+    li.QuadPart = Data.wSize - sizeof(Data);
+    hr = pOldStream->Seek(li, STREAM_SEEK_CUR, NULL);
+    if (FAILED(hr))
+        return NULL;
+
+    // Get the size
+    ULARGE_INTEGER uli;
+    hr = IStream_Size(pOldStream, &uli);
+    if (FAILED(hr))
+        return NULL;
+
+    // Create new stream and attach
+    CComPtr<IStream> pNewStream;
+    pNewStream.Attach(SHCreateMemStream(NULL, 0));
+    if (!pNewStream)
+        return NULL;
+
+    // Subtract Data.wSize from the size
+    uli.QuadPart -= Data.wSize;
+
+    // Copy to pNewStream
+    hr = pOldStream->CopyTo(pNewStream, uli, NULL, NULL);
+    if (FAILED(hr))
+        return NULL;
+
+    li.QuadPart = 0;
+    pNewStream->Seek(li, STREAM_SEEK_SET, NULL);
+
+    return pNewStream.Detach();
+}
+
+IStream* CDesktopUpgradePropertyBag::_GetOldDesktopViewStream()
+{
+    HKEY hKey = SHGetShellKey(SHKEY_Root_HKCU, L"Streams\\Desktop", FALSE);
+    if (!hKey)
+        return NULL;
+
+    CComPtr<IStream> pOldStream;
+    if (!_AlreadyUpgraded(hKey))
+    {
+        pOldStream.Attach(SHOpenRegStream2W(hKey, NULL, L"ViewView2", 0));
+        if (pOldStream)
+        {
+            ULARGE_INTEGER uli;
+            HRESULT hr = IStream_Size(pOldStream, &uli);
+            if (SUCCEEDED(hr) && !uli.QuadPart)
+                pOldStream.Release();
+        }
+
+        if (!pOldStream)
+            pOldStream.Attach(SHOpenRegStream2W(hKey, NULL, L"ViewView", 0));
+
+        _MarkAsUpgraded(hKey);
+    }
+
+    ::RegCloseKey(hKey);
+    return pOldStream.Detach();
+}
+
+HRESULT CDesktopUpgradePropertyBag::_ReadItemPositions(VARIANT *pvari)
+{
+    CComPtr<IStream> pOldStream;
+    pOldStream.Attach(_GetOldDesktopViewStream());
+    if (!pOldStream)
+        return E_FAIL;
+
+    HRESULT hr = E_FAIL;
+    IStream *pNewStream = _NewStreamFromOld(pOldStream);
+    if (pNewStream)
+    {
+        V_UNKNOWN(pvari) = pNewStream;
+        V_VT(pvari) = VT_UNKNOWN;
+        hr = S_OK;
+    }
+
+    return hr;
+}
+
+STDMETHODIMP
+CDesktopUpgradePropertyBag::Read(
+    _In_z_ LPCWSTR pszPropName,
+    _Inout_ VARIANT *pvari,
+    _Inout_opt_ IErrorLog *pErrorLog)
+{
+    UNREFERENCED_PARAMETER(pErrorLog);
+
+    VARTYPE vt = V_VT(pvari);
```

Comment:
Why is it necessary to cache the vartype there and not "later" (or just use `V_VT(pvari)` in the VariantChangeTypeForRead call) ? Is it because the vartype may change during the function calls below?

Reply:
`VariantChangeTypeForRead` is (`VariantChangeType` + `BSTR` reinterpretation).
We don't use `BSTR` here, so `VariantChangeType` is enough.
Both `_ReadFlags` and `_ReadItemPositions` can change the var type (`V_VT(pvari)`).
Here `vt` will be the target var type.